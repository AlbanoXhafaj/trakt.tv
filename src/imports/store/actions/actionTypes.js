export const fetchActionTypes = {
  DATA_FETCHED: "DATA_FETCHED",
  FETCH_NEW_RESOURCE: "FETCH_NEW_RESOURCE",
  FETCHING_ERROR: "FETCHING_ERROR",
  START_FETCHING: "START_FETCHING"
}

export const filtersActionTypes = {
  SET_SORT_FILTER: "SET_SORT_FILTER",
  SET_PERIOD_FILTER: "SET_PERIOD_FILTER",
  SET_GRID_SIZE_FILTER: "SET_GRID_SIZE_FILTER",
  SET_SEARCH_FILTER: "SET_SEARCH_FILTER"
};

export const paginationActionTypes = {
  SWITCH_PAGE: "SWITCH_PAGE",
  LOAD_NEW_PAGE_INDEXES: "LOAD_NEW_PAGE_INDEXES",
  SHOW_DEFAULT_PAGE: "SHOW_DEFAULT_PAGE"
};
