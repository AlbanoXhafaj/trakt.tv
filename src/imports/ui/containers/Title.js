import React from "react";
import PropTypes from 'prop-types';

//It accepts the title of page as children
const Title = ({ children }) => (
  <div id="header">
    <h1>{ children }</h1>
  </div>
);

Title.propTypes = {
  children: PropTypes.string.isRequired
};

export default Title;
