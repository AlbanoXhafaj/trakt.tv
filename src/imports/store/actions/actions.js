import { fetchActionTypes , filtersActionTypes , paginationActionTypes } from "./actionTypes";

// ACTIONS FOR FETCH REDUCER

//Fetch promise resolves successfully
export const  dataFetched = ( pagesNumber , shows ) =>({
  type: fetchActionTypes.DATA_FETCHED,
  pagesNumber,
  shows
});

//Fetch promise is rejected
export const fetchingError = () => ({
  type: fetchActionTypes.FETCHING_ERROR
});

//Fetch another resource
export const fetchNewResource = () => ({
  type: fetchActionTypes.FETCH_NEW_RESOURCE
});

//Fetching started
export const startFetching = () =>({
  type: fetchActionTypes.START_FETCHING
});

// ACTIONS FOR PAGINATION REDUCER

//Switch page index like page: "2"
export const switchPage = (pageIndex) => ({
  type: paginationActionTypes.SWITCH_PAGE,
  pageIndex
});

//Load more indexes in pagination panel
export const loadNewPageIndexes = startIndex => ({
  type: paginationActionTypes.LOAD_NEW_PAGE_INDEXES,
  startIndex
});

export const showDefaultPage = () => ({
  type: paginationActionTypes.SHOW_DEFAULT_PAGE
});

// ACTIONS FOR FILTERS REDUCER

//set new sort filter
export const setSortFilter = sortFilter => ({
  type: filtersActionTypes.SET_SORT_FILTER,
  sortFilter
});

//set new period filter
export const setPeriodFilter = periodFilter => ({
  type: filtersActionTypes.SET_PERIOD_FILTER,
  periodFilter
});

//set new grid size filter
export const setGridSizeFilter = gridSizeFilter => ({
  type: filtersActionTypes.SET_GRID_SIZE_FILTER,
  gridSizeFilter
});

//set new searching text
export const setSearchFilter = searchFilter => ({
  type: filtersActionTypes.SET_SEARCH_FILTER,
  searchFilter
});
