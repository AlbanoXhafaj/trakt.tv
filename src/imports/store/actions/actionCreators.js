import  traktTvApi from "api/TraktTvAPI";
import * as actions from "./actions";
import { filters } from "reducers/FiltersReducer";

/*Sets FetchStatus to fetching , starts fetching from the api
  addsdispatches corresponding action when resolves or rejects */
export const fetchData = ( url , pageIndex , nrItems ) => (
  dispatch => {
    dispatch(actions.startFetching());

  return traktTvApi.getShows( url , pageIndex , nrItems ).then(
      ({ pagesNumber , shows }) => dispatch(actions.dataFetched( pagesNumber , shows )),
      (error) => dispatch( actions.fetchingError() )
    );
  }
);

//Loads descending indexes by subtracting the start
export const loadLeftPaginationIndexes = ( startIndex , displayedIndexes ) => (
  dispatch => {                                  //example
    let newStartIndex =                          //display 10 indexes, number of pages 25
      startIndex - displayedIndexes > 1          //current index a)5 , b)18
        ? startIndex - displayedIndexes          // a) 5-10 < 1 return 1
        : 1;                                     // b) 18-10 > 1 return 8

    return dispatch( actions.loadNewPageIndexes(newStartIndex) );
  }
);

//Loads ascending indexes by subtracting the start
/*example
  display 10 indexes , number of pages 25
  In order always to display 10 last indexes the inequation is:
  startIndex + displayedIndexes < pagesNumber - displayedIndexes that is equal to
  startIndex + 2 * displayedIndexes < pagesNumber

  a) 7, b)3
  a) 7 + 10 < 25 - 10 <=> 7 + 20 < 25 return 25 - 10;
  b) 3 + 10 < 25 - 10 <=> 3 + 20 < 25 return 3 + 10;
*/
export const loadRightPaginationIndexes = ( startIndex , pagesNumber, displayedIndexes ) => (
  dispatch => {
    let newStartIndex =
      startIndex + 2 * displayedIndexes < pagesNumber
        ? startIndex + displayedIndexes
        : pagesNumber - displayedIndexes > 0
            ? pagesNumber - displayedIndexes
            : 1;

    return dispatch( actions.loadNewPageIndexes(newStartIndex) );
  }
)

//Switches page and loads the new page
export const switchPage = (newPageIndex) =>(
  dispatch => {
    dispatch(actions.switchPage( newPageIndex ));
    dispatch(actions.fetchNewResource());
  }
);

//Sets new filter and load new resource
export const setSortFilter = ( index ) => (
  dispatch => {
    const filter = filters.sortFilters.getFilter(index);
    dispatch(actions.setSortFilter(filter));
    dispatch(actions.showDefaultPage());
    dispatch(actions.fetchNewResource());
  }
)

//Sets new filter and load new resource
export const setPeriodFilter = ( index ) => (
  dispatch => {
    const filter = filters.periodFilters.getFilter(index);
    dispatch(actions.setPeriodFilter(filter));
    dispatch(actions.showDefaultPage());
    dispatch(actions.fetchNewResource());
  }
)

//Sets new filter and load new resource
export const setGridSizeFilter = ( index ) => (
  dispatch => {
    const filter = filters.gridSizeFilters.getFilter(index);
    dispatch(actions.setGridSizeFilter(filter));
    dispatch(actions.showDefaultPage());
    dispatch(actions.fetchNewResource());
  }
)

//Sets new filter and load new resource
export const setSearchFilter = ( searchText ) => (
  dispatch => {
    const filter = filters.searchFilter.getFilter(searchText); // trim() is optional
    dispatch(actions.setSearchFilter(filter));
    dispatch(actions.showDefaultPage());
    dispatch(actions.fetchNewResource());
  }
);
