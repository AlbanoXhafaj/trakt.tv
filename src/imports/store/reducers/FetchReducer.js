import { fetchActionTypes } from "../actions/actionTypes";

export const fetchStatus = {
  newFetch: "newFetch",
  fetching: "fetching",
  error: "error",
  complete: "complete"
};

export const initialState = {
  status: fetchStatus.newFetch,
  data: {
    pagesNumber: 0,
    shows: []
  }
}

const fetchReducer = (state = initialState, action) => {
  switch(action.type){
    case fetchActionTypes.DATA_FETCHED:   //when the data is fetched and page load
    return {
      status: fetchStatus.complete,
      data:{
        pagesNumber: action.pagesNumber,
        shows: action.shows
      }
    }
    case fetchActionTypes.FETCHING_ERROR: //when couldn't fetch data from api from an unhandable reason
    return {
      ...state,
      status: fetchStatus.error
    }
    case fetchActionTypes.START_FETCHING:
    return {
      ...state,
      status: fetchStatus.fetching
    }
    case fetchActionTypes.FETCH_NEW_RESOURCE:   // initial state that idicates loading from the api
    return initialState;
    default:
    return state;
  }
}

export default fetchReducer;
