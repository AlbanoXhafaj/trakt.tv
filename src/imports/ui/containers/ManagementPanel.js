import React from "react";
import PropTypes from 'prop-types';

import SelectList from "containers/SelectList";
import SearchPanel from "containers/SearchPanel";

import * as actionCreators from "actions/actionCreators";
import { filters } from "reducers/FiltersReducer";
import * as selectors from "store/stateSelectors";
import { connect } from "react-redux";

//Creates 2 SelectLists, 1 SearchPanel and if needed 1 extra Select List for the period filters
export const _ManagementPanel = ({ sortFilter, searchText , setSortFilter , setPeriodFilter, setGridSizeFilter , setSearchFilter }) => (
  <fieldset>
    <legend>Management Panel</legend>
    <SelectList
      title="Sort"
      options = { filters.sortFilters.getValues()  }
      onChange = { setSortFilter }
    />
    {
      sortFilter.needPeriodParameter
        ? <SelectList
            title="Period"
            options = { filters.periodFilters.getValues() }
            onChange = { setPeriodFilter }
          />
        : null
    }
    <SearchPanel onChange = { setSearchFilter } text = { searchText }/>
    <SelectList
      title="Grid Size"
      options = { filters.gridSizeFilters.getValues() }
      onChange = { setGridSizeFilter }
    />
  </fieldset>
);

export const mapStateToProps = (state) =>({
  sortFilter: selectors.getSortFilter(state),
  searchText: selectors.getSearchText(state)
});

export const mapDispatchToProps = {
  setSortFilter: actionCreators.setSortFilter,
  setPeriodFilter: actionCreators.setPeriodFilter,
  setGridSizeFilter: actionCreators.setGridSizeFilter,
  setSearchFilter: actionCreators.setSearchFilter
};

const ManagementPanel = connect(mapStateToProps,mapDispatchToProps)(_ManagementPanel);

ManagementPanel.propTypes = {
  sortFilter: PropTypes.exact({
    url: PropTypes.string,
    value: PropTypes.string,
    needPeriodParameter: PropTypes.bool
  }).isRequired,
  searchText: PropTypes.string.isRequired,
  searchFilter: PropTypes.func.isRequired,
  setPeriodFilter: PropTypes.func.isRequired,
  setGridSizeFilter: PropTypes.func.isRequired,
  setSearchFilter: PropTypes.func.isRequired
};

export default ManagementPanel;
