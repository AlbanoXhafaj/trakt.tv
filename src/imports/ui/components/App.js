import React from "react";
import PropTypes from 'prop-types';
import { connect } from "react-redux";

import Title from "containers/Title";
import ErrorSign from "containers/ErrorSign";
import LoadingProcess from "containers/LoadingProcess";
import InfoMessage from "containers/InfoMessage";
import Table from "containers/Table";
import PaginationPanel from "containers/PaginationPanel";
import ManagementPanel from "containers/ManagementPanel";

import * as actionCreators from "actions/actionCreators";
import * as selectors from "store/stateSelectors";
import { fetchStatus  as FetchStatus } from "reducers/FetchReducer";


//A component that checks for fetched every time after render,
//in order to dispatch actions after renders and while rendering,
//because may cause an unexpected behaviour from the application
export class _App extends React.Component{

  componentDidUpdate(){
    shouldFetch(this.props);
  }

  componentDidMount(){
   shouldFetch(this.props);
  }

  render(){
    return(
      <div className="app" >
        <Title>{ this.props.title }</Title>
        { getBodyContainer(this.props) }
        <ManagementPanel />
      </div>
    )
  }
}

//Controls if we should fetch new data, is so we fetch
export const shouldFetch = ( {fetchStatus , pageIndex , url , nrItems , fetchData } ) => {
  if( fetchStatus === FetchStatus.newFetch ) fetchData( url , pageIndex , nrItems );
}

//Extracts data needed for the display from the props
export const getBodyContainer = ( { fetchStatus , shows } ) => {

  const NoShows = !shows || shows.length === 0 ? true : false;

  if(fetchStatus === FetchStatus.fetching )  return <LoadingProcess />;
  if( fetchStatus === FetchStatus.error ) return <ErrorSign />;
  if( NoShows ) return <InfoMessage>No shows found.</InfoMessage>;

   return (
    <div className="center" >
      <Table />
      <PaginationPanel />
    </div>
  )
}

const mapStateToProps = (state) => ({
  fetchStatus: selectors.getFetchStatus(state),
  shows: selectors.getShows(state),
  pageIndex: selectors.getPageIndex(state),
  nrItems: selectors.getNrItems(state),
  url: selectors.getUrl(state),
  title: selectors.getTitle(state)
});


const mapDispatchToProps = {
  fetchData: actionCreators.fetchData
};

const App = connect( mapStateToProps , mapDispatchToProps)(_App);

App.propTypes = {
  fetchStatus: PropTypes.string.isRequired,
  shows: PropTypes.arrayOf(PropTypes.object).isRequired,
  pageIndex: PropTypes.number.isRequired,
  nrItems: PropTypes.number.isRequired,
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  fetchData: PropTypes.func.isRequired
};

export default App;
