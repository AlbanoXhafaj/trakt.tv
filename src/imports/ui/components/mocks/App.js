import React from "react";
import Title from "containers/Title";
import ManagementPanel from "containers/ManagementPanel";
import { fetchStatus } from "reducers/FetchReducer";
import * as realApp from "../App";

//Mocks the real function which returns different containers depending on the props
const mockGetBodyContainer = () => <div className="bodyContainer"></div>;

//A mock component in order to test the lifecycle events
class App extends React.Component{

  mockShouldFetch(){
    realApp.shouldFetch(this.props);
  };

  componentDidUpdate(){
    this.mockShouldFetch();
  }

  componentDidMount(){
   this.mockShouldFetch();
  }

  render(){
    return(
      <div className="app" >
        <Title>{ this.props.title }</Title>
        { mockGetBodyContainer() }
        <ManagementPanel />
      </div>
    )
  }
}

//Export the original function of the component for testing
export const getBodyContainer = (props) => realApp.getBodyContainer(props);

export default App;
