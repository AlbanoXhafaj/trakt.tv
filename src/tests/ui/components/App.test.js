import "../../setups/setup_enzyme.js";
import React from "react";
import { shallow , mount } from "enzyme";
import toJSON from 'enzyme-to-json';

import { fetchStatus } from "reducers/FetchReducer";

jest.mock("containers/ManagementPanel" , () => () => <div>ManagementPanel</div>);

import App from "components/App";
import Title from "containers/Title";
import ManagementPanel from "containers/ManagementPanel";
import * as actionCreators from "actions/actionCreators";

actionCreators.fetchData = jest.fn();

describe("App Component", () => {

  const url = "www.works.com";
  const pageIndex = 1;
  const nrItems = 10;

  const initialProps = {
    title: "Shows",
    fetchStatus: fetchStatus.newFetch,
    fetchData: actionCreators.fetchData,
    url,
    pageIndex,
    nrItems
  };

  const wrapper = shallow( <App  { ...initialProps } />);

  const instance = wrapper.instance();

  it("renders without problem",() => {
    expect(wrapper.length).toBe(1);
  })

  it("renders the title", () => {
    expect(wrapper.find(Title).length).toBe(1);
  });

  it("render the body container" ,() => {
    expect(wrapper.find("div.bodyContainer").length).toBe(1);
  });

  it("renders the ManagementPanel", () => {
    expect(wrapper.find(ManagementPanel).length).toBe(1);
  });

  it("check for new fetch on componentDidMount and fetch" ,() => {
    expect(instance.componentDidMount.calledOnce);
    expect(actionCreators.fetchData).toHaveBeenCalledWith(url,pageIndex,nrItems);
  });

  it("checks for the new fetch on componentDidUpdate" ,() => {
    Object.keys(fetchStatus).forEach( key => {
      actionCreators.fetchData.mockClear();

      wrapper.setProps({ fetchStatus: fetchStatus[key] });

      expect(instance.componentDidUpdate.called);

      if(fetchStatus[key] === fetchStatus.newFetch)
        expect(actionCreators.fetchData).toHaveBeenCalledWith(url,pageIndex,nrItems);
      else
        expect(actionCreators.fetchData).not.toHaveBeenCalled();
    })
  });

  it("matches to the snapshot" , () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
