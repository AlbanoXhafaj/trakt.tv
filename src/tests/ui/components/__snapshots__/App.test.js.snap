// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches to the snapshot 1`] = `
<div
  className="app"
>
  <Title>
    Shows
  </Title>
  <div
    className="bodyContainer"
  />
  <Component />
</div>
`;
